hazelcast/src/main/java/com/hazelcast/jet/pipeline/RemoteMapSourceBuilder.java

```diff
@@ -0,0 +1,176 @@
+ * available when using {@link Projections#singleAttribute} and {@link
+ * Projections#multiAttribute}) to create your projection instance and
+ * using the {@link Predicates} factory or
+ * {@link PredicateBuilder PredicateBuilder} to create
+ * the predicate. In this case Jet can test the predicate and apply the
+ * projection without deserializing the whole object.
+ * <p>
+ * Due to the current limitations in the way Jet reads the map it can't use
+ * any indexes on the map. It will always scan the map in full.
+ * <p>
+ * The source does not save any state to snapshot. If the job is restarted,
+ * it will re-emit all entries.
+ * <p>
+ * If the {@code IMap} is modified while being read, or if there is a
+ * cluster topology change (triggering data migration), the source may miss
+ * and/or duplicate some entries. If we detect a topology change, the job
+ * will fail, but the detection is only on a best-effort basis - we might
+ * still give incorrect results without reporting a failure. Concurrent
+ * mutation is not detected at all.
+ * <p>
+ * The default local parallelism for this processor is 1.
+ * <p>
+ * <h4>Predicate/projection class requirements</h4>
+ *
+ * The classes implementing {@code predicate} and {@code projection} need
+ * to be available on the remote cluster's classpath or loaded using
+ * <em>Hazelcast User Code Deployment</em>. It's not enough to add them to
+ * the job classpath in {@link JobConfig}. The same is true for the class
+ * of the objects stored in the map itself. If you cannot meet these
+ * conditions, do not add predicate and projection and add a
+ * subsequent {@link GeneralStage#map map} or {@link GeneralStage#filter
+ * filter} stage.
+ *<p>
+ * @param <K> the type of the key in the map
+ * @param <V> the type of the value in the map
+ * @param <T> the type of the emitted items
+ * @since 5.4
+ */
+public class RemoteMapSourceBuilder<K, V, T> {
+
+    private final String mapName;
+    private String dataConnectionName;
+    private ClientConfig clientConfig;
+    private Predicate<K, V> predicate;
+    private Projection<? super Entry<K, V>, ? extends T> projection;
+
+    RemoteMapSourceBuilder(String mapName) {
+        requireNonNull(mapName, "mapName can not be null");
+        this.mapName = mapName;
+    }
+
+    /**
+     * Set the data connection name to use to connect to the remote cluster.
+     * The data connection must be of
+     * {@link com.hazelcast.dataconnection.HazelcastDataConnection} type.
+     * <p>
+     * One of {@link #dataConnectionName} or {@link #clientConfig} is required
+     * to build the source. If both are provided the data connection takes
+     * precedence.
+     *
+     * @param dataConnectionName name of the data connection
+     * @return this builder
+     */
+    public RemoteMapSourceBuilder<K, V, T> dataConnectionName(@Nonnull String dataConnectionName) {
+        this.dataConnectionName = requireNonNull(dataConnectionName);
```

Comment:
please add second argument with message - otherwise NPE will be not very useful

Reply:
rev : 680737fe0c040b92fd744f238ca4d38b1f8f926f
Performing null check and serializable checks independently with different messages